tests/harness/src/case.rs

```diff
@@ -0,0 +1,153 @@
+use std::{fmt::Display, path::Path};
+
+use anyhow::{Context, Result};
+use ceresdb_client_rs::{
+    client::{Client, RpcContext},
+    model::{display::CsvFormatter, request::QueryRequest, QueriedRows},
+    DbClient,
+};
+use tokio::{
+    fs::File,
+    io::{AsyncBufReadExt, AsyncWrite, AsyncWriteExt, BufReader},
+};
+
+const INTERCEPTOR_PREFIX: &'static str = "-- CERESDB";
+const COMMENT_PREFIX: &'static str = "--";
+
+pub struct TestCase {
+    name: String,
+    queries: Vec<Query>,
+}
+
+impl TestCase {
+    pub async fn from_file<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let file = File::open(path.as_ref())
+            .await
+            .with_context(|| format!("Cannot open file {:?}", path.as_ref()))?;
+
+        let mut queries = vec![];
+        let mut query = Query::default();
+
+        let mut lines = BufReader::new(file).lines();
+        while let Some(line) = lines
+            .next_line()
+            .await
+            .with_context(|| format!("Failed to read from file {:?}", path.as_ref()))?
+        {
+            // intercept command start with INTERCEPTOR_PREFIX
+            if line.starts_with(INTERCEPTOR_PREFIX) {
+                query.push_interceptor(line);
+                continue;
+            }
+
+            // ignore comment and empty line
+            if line.starts_with(COMMENT_PREFIX) || line.is_empty() {
+                continue;
+            }
+
+            query.append_query_line(&line);
+
+            // SQL statement ends with ';'
+            if line.ends_with(';') {
+                queries.push(query);
+                query = Query::default();
+            }
+        }
+
+        Ok(Self {
+            name: path.as_ref().to_str().unwrap().to_string(),
+            queries,
+        })
+    }
+
+    pub async fn execute<W>(&self, client: &Client, writer: &mut W) -> Result<()>
+    where
+        W: AsyncWrite + Unpin,
+    {
+        for query in &self.queries {
+            query
+                .execute(client, writer)
+                .await
+                .with_context(|| format!("Error while executing {:?}", query.query_lines))?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Display for TestCase {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.name)
+    }
+}
+
+#[derive(Default)]
+struct Query {
+    query_lines: Vec<String>,
+    interceptors: Vec<String>,
+}
+
+impl Query {
+    fn push_interceptor(&mut self, post_process: String) {
+        self.interceptors.push(post_process);
+    }
+
+    fn append_query_line(&mut self, line: &str) {
+        self.query_lines.push(line.to_string());
+    }
+
+    async fn execute<W>(&self, client: &Client, writer: &mut W) -> Result<()>
```

Comment:
This execute only test via gRPC, should we also support MySQL?

Reply:
One is enough I think. The difference is limited to the protocol level. Both protocols are using (and should use) the same underlying interfaces.

Reply:
> Both protocols are using (and should use) the same underlying interfaces.

Yes, we only test the frontend layer of MySQL, if we recommend MySQL protocol in production, then it's best this test cover that.

Reply:
The protocol frontend is only used for communicating with server but not the target of those test cases. Protocols should have their own suits to guarantee the behavior.